Define key exporters (#148)
ietf-wg-webtrans/draft-ietf-webtrans-http3: draft-ietf-webtrans-http3.md

streams.  The signal is intended for the application using WebTransport, which
is expected to attempt to gracefully terminate the session as soon as possible.

## Use of Keying Material Exporters

WebTransport over HTTP/3 supports the use of TLS keying material exporters
{{!RFC5705}}.  Since the underlying QUIC connection may be shared by multiple
WebTransport sessions, WebTransport defines its own mechanism for deriving a
TLS exporter that separates keying material for different sessions.  If the
user requests an exporter for a given WebTransport session with a specified

> label and context, the resulting exporter SHALL be a TLS exporter as defiend in

Reviewer: ```suggestion
label and context, the resulting exporter SHALL be a TLS exporter as defined in
```

> {{RFC5705, Section 4}} with the label set to "EXPORTER-WebTransport" and the

Reviewer: ```suggestion
{{RFC8446, Section 7.5}} with the label set to "EXPORTER-WebTransport" and the
```

context set to the serialization of the "WebTransport Exporter Context" struct
as defined below.

~~~
WebTransport Exporter Context {
  WebTransport Session ID (64),

>   WebTransport Application-Supplied Exporter Label Length (32),

Reviewer: 32 bits is monstrously long.  TLS exporters only permit 0..255, so I would strongly suggest that you change these:

```suggestion
  WebTransport Application-Supplied Exporter Label Length (8),
```

You might also note that there is a further limitation on the total size imposed by TLS.

  WebTransport Application-Supplied Exporter Label (..),
  WebTransport Application-Supplied Exporter Context Length (32),
  WebTransport Application-Supplied Exporter Context (..)
}
~~~

> If the user does not supply any context, it is treated as the context
> with zero length.

Reviewer: ```suggestion
A TLS exporter API might permit the context field to be omitted. In this case,
as with TLS 1.3, the WebTransport Application-Supplied Exporter Context
becomes zero-length if omitted.
```

streams.  The signal is intended for the application using WebTransport, which
is expected to attempt to gracefully terminate the session as soon as possible.

## Use of Keying Material Exporters

WebTransport over HTTP/3 supports the use of TLS keying material exporters

> {{!RFC5705}}.  Since the underlying QUIC connection may be shared by multiple

Reviewer: ```suggestion
{{!RFC8446, Section 7.5}}.  Since the underlying QUIC connection may be shared by multiple
```

WebTransport sessions, WebTransport defines its own mechanism for deriving a
TLS exporter that separates keying material for different sessions.  If the
user requests an exporter for a given WebTransport session with a specified
label and context, the resulting exporter SHALL be a TLS exporter as defined in

> {{RFC8446, Section 7.5}} with the label set to "EXPORTER-WebTransport" and the

Reviewer: ```suggestion
{{!RFC8446, Section 7.5}} with the label set to "EXPORTER-WebTransport" and the
```

> {{RFC8446, Section 7.5}} with the label set to "EXPORTER-WebTransport" and the

Reviewer 2: ```suggestion
{{!RFC8446, Section 7.5}} with the label set to "EXPORTER-WebTransport" and the
```

context set to the serialization of the "WebTransport Exporter Context" struct
as defined below.

~~~
WebTransport Exporter Context {
  WebTransport Session ID (64),
  WebTransport Application-Supplied Exporter Label Length (8),

>   WebTransport Application-Supplied Exporter Label (..),

Reviewer: ```suggestion
  WebTransport Application-Supplied Exporter Label (8..),
```

It probably makes sense to set a minimum.

>   WebTransport Application-Supplied Exporter Context Length (32),

Reviewer: ```suggestion
  WebTransport Application-Supplied Exporter Context Length (8),
```

Reviewer: Note: it is not sufficient to just merge suggestions.

Author: Ah, I misread that as implying that only the label was subject to length restrictions.

  WebTransport Application-Supplied Exporter Context (..)
}
~~~

> 

Reviewer: ```suggestion

Note that the TLS exporter API requires that labels are between 1 and 249 bytes in length and contexts are between 0 and 255 bytes.  Because this structure is included in the context field of a TLS exporter, the combined length of labels and contexts needs to be 245 bytes or shorter.

```

Reviewer: As Victor notes, this limit doesn't really need to be enforced due to the hashing limit.
